src/cinematic_waypoints/cinematic_controller.py

```diff
@@ -1,119 +1,134 @@
+
+        best_bb = None
+        min_error = math.inf
+        for bb in bounding_boxes:
+            # Use Euclidean distance of dimensions and centroid to compute the "error"
+            bb_dimensions = np.asarray(bb.get_dimensions())
+            dimension_error = numpy.linalg.norm(bb_dimensions - np.asarray(self.smoothed_bounding_box.get_dimensions()))
+            bb_centroid = np.asarray(bb.get_centroid())
+            centroid_error = numpy.linalg.norm(bb_centroid - np.asarray(self.smoothed_bounding_box.get_centroid()))
+
+            error = self.bb_match_dimension_importance * dimension_error + centroid_importance * centroid_error
+            if error < min_error:
+                min_error = error
+                best_bb = bb
+
+        return best_bb
+
+    # Returns the bounding box from the list provided with the largest area.
+    @staticmethod
+    def identify_largest_bb(bounding_boxes):
+        largest_bb = None
+        largest_area = -1
+        for bb in bounding_boxes:
+            bb_area = bb.get_area()
+            if bb_area > largest_area:
+                largest_area = bb_area
+                largest_bb = bb
+        return largest_bb
+
+    def update_latest_drone_state(self, drone_state):
+        self.latest_drone_state = drone_state
+
+    def update_cinematic_waypoints(self, cinematic_waypoints):
+        self.cinematic_waypoints = cinematic_waypoints
+
+    # Uses a geometric low-pass filter to smooth out human bounding boxes
+    def compute_low_pass_filter_bb(self):
+        latest_width, latest_height = self.latest_bounding_box.get_dimensions()
+        latest_x, latest_y = self.latest_bounding_box.get_centroid()
+
+        smoothed_width, smoothed_height = self.smoothed_bounding_box.get_dimensions()
+        smoothed_x, smoothed_y = self.smoothed_bounding_box.get_centroid()
+
+        new_width = smoothed_width * self.bb_filter_gamma + latest_width * (1 - self.bb_filter_gamma)
+        new_height = smoothed_height * self.bb_filter_gamma + latest_height * (1 - self.bb_filter_gamma)
+        new_x = smoothed_x * self.bb_filter_gamma + latest_x * (1 - self.bb_filter_gamma)
+        new_y = smoothed_y * self.bb_filter_gamma + latest_y * (1 - self.bb_filter_gamma)
+        return BoundingBox((new_width, new_height), (new_x, new_y))
+
+    def generate_waypoints(self):
+        if self.smoothed_bounding_box is None or self.latest_drone_state is None:
+            print("Warning: asked to generate waypoints, but some of the current states are None.")
+        if self.cinematic_waypoints==None:
+            return self.waypoint_generator.generate_waypoints(self.smoothed_bounding_box, self.latest_drone_state)
+        else:
+            waypoint = self.cinematic_waypoints[0]
+            check_x,check_y,check_z = self.latest_drone_state-waypoint
+            if check_x < self.margin and check_y < self.margin and check_z < self.margin:
+                del(self.cinematic_waypoints[0])

```

Comment:
Wow. I didn't know about the del function. It seems like the right method for the job. Are those parentheses the right syntax? It could be a python 2 vs. python 3 thing, but here they don't seem to use parentheses: https://stackoverflow.com/questions/11520492/difference-between-del-remove-and-pop-on-lists